src/Drivers/Digipay/Digipay.php

```diff
@@ -220,6 +228,111 @@ protected function oauth()
         return $body['access_token'];
     }
 
+    /**
+     * @see https://docs.mydigipay.com/upg.html#_purchase_reverse
+     * @throws PurchaseFailedException
+     */
+    public function reverse()
+    {
+        if (is_null($digipayTicketType = $this->invoice->getDetail('type'))) {
+            throw new PurchaseFailedException('"type" is required for this method.');
+        }
+
+        if (is_null($trackingCode = $this->invoice->getDetail('trackingCode'))) {
+            throw new PurchaseFailedException('"trackingCode" is required for this method.');
+        }
+
+        $data = [
+            'trackingCode' => $trackingCode,
+            'providerId'    => $this->invoice->getTransactionId() ?? $this->invoice->getDetail('providerId') ?? $this->invoice->getUuid(),
+        ];
+
+        $response = $this
+            ->client
+            ->request(
+                'POST',
+                $this->settings->apiPaymentUrl . self::REVERSE_URL,
+                [
+                    RequestOptions::BODY        => json_encode($data),
+                    RequestOptions::QUERY       => ['type' => $digipayTicketType],
+                    RequestOptions::HEADERS     => [
+                        'Content-Type'    => 'application/json;charset=UTF-8',
+                        'Authorization'   => 'Bearer ' . $this->oauthToken,
+                    ],
+                    RequestOptions::HTTP_ERRORS => false,
+                ]
+            );
+
+        $body = json_decode($response->getBody()->getContents(), true);
+
+        if ($response->getStatusCode() != 200 || $body['result']['code'] != 0) {
+            $message = $body['result']['message'] ?? 'خطا در هنگام درخواست برای برگشت وجه رخ داده است.';
+            throw new InvalidPaymentException($message, (int) $response->getStatusCode());
+        }
+
+        return $body;
+    }
+
+    /**
+     * @see https://docs.mydigipay.com/upg.html#_purchase_delivery
+     * @throws PurchaseFailedException
+     */
+    public function deliver()
+    {
+        $details = ['type', 'invoiceNumber', 'deliveryDate', 'trackingCode', 'products'];
+        foreach ($details as $detail) {
+            if (empty($value = $this->invoice->getDetail($detail))) {
+                throw new PurchaseFailedException("\"$detail\" is required for this method.");
+            }
+            // define variable with the same name as the value of $detail
+            $$detail = $value;
```

Comment:
this is not a good practice and it makes debugging hard.

Reply:
update this section.